synapse/storage/databases/main/events_bg_updates.py

```diff
@@ -932,3 +937,81 @@ def _calculate_chain_cover_txn(
             processed_count=count,
             finished_room_map=finished_rooms,
         )
+
+    async def _purged_chain_cover_index(self, progress: dict, batch_size: int) -> int:
+        """
+        A background updates that iterates over the chain cover and deletes the
+        chain cover for events that have been purged.
+
+        This may be due to fully purging a room or via setting a retention policy.
+        """
+        current_event_id = progress.get("current_event_id", "")
+
+        def purged_chain_cover_txn(txn) -> int:
+            sql = """
+                SELECT event_id, chain_id, sequence_number FROM event_auth_chains
+                WHERE event_id > ? ORDER BY event_id ASC LIMIT ?
+                """
+            txn.execute(sql, (current_event_id, batch_size))
+
+            rows = txn.fetchall()
+            if not rows:
+                return 0
+
+            # The event IDs and chain IDs / sequence numbers where the event has
+            # been purged.
+            unreferenced_event_ids = []
+            unreferenced_chain_id_tuples = []
+            event_id = ""
+            for row in rows:
+                event_id = row[0]
+
+                txn.execute(
+                    """
+                    SELECT event_id FROM event_json WHERE event_id = ?
```

Comment:
I'd combine this with the query above and do:

```sql
SELECT event_id, chain_id, sequence_number, e.event_id
FROM event_auth_chains
LEFT JOIN events AS e USING (event_id)
WHERE event_id > ? ORDER BY event_id ASC LIMIT ?
```

The last column will then be `NULL` if we don't have an entry in the events table.